core/src/main/java/com/opensymphony/xwork2/ognl/accessor/CompoundRootAccessor.java

```diff
@@ -233,24 +234,22 @@ public Object callMethod(Map context, Object target, String name, Object[] objec
 
             if ((argTypes == null) || !invalidMethods.containsKey(mc)) {
                 try {
-                    Object value = OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
-
-                    if (value != null) {
-                        return value;
-                    }
+                    return OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
                 } catch (OgnlException e) {
                     // try the next one
-                    Throwable reason = e.getReason();
+                    reason = e.getReason();
 
-                    if (!context.containsKey(OgnlValueStack.THROW_EXCEPTION_ON_FAILURE) && (mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
+                    if ((mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
                         invalidMethods.put(mc, Boolean.TRUE);
-                    } else if (reason != null) {
```

Comment:
Is it certain that dropping the "if (reason != null)" check as part of the decision whether to throw the exception or not won't cause an issue ?

It looks like the old code would "skip" throwing an Exception in that case, no matter what `THROW_EXCEPTION_ON_FAILURE` is set to ?

Reply:
As @qqu0127 mentioned, it seems `reason != null` tries to distinguish between `property not found` and `user app` exception but actually it doesn't work. As far as I could see and debug, OGNL always fill it here
 
https://github.com/jkuhnert/ognl/blob/b6e58c4db776beb20e20d5792227d1da7b4355e0/src/main/java/ognl/OgnlRuntime.java#L1529

in both `NoSuchMethodException` (i.e. property not found) and `InvocationTargetException` (i.e. user app exception).

I realized that this is an old issue, please see this test : https://github.com/apache/struts/blob/d4dd3386cc02833fa7113a496092b2e09a53c9e4/core/src/test/java/com/opensymphony/xwork2/ognl/OgnlValueStackTest.java#L958

Reply:
Oh, you are right. `reason != null` itself is not a complete check. Actually, I fixed this in my workplace integrated with our logging system internally, based on 2.5.17.  And it classifies the exceptions correctly. FYI, we modified `ValueStack`. (we also changed `CompoundRootAccessor`, but differently, so not sure if it works here)
```
  @Override
  public Object handleOgnlException(String expr, boolean throwExceptionOnFailure, OgnlException e) {
    Object ret = findInContext(expr);
    if (ret != null) {
      return ret;
    }
    String errMsg = null;
    if (shouldLogWarning()) {
      if (e.getCause() == null || e.getCause() instanceof NoSuchMethodException) {
        errMsg = String.format("Could not evaluate expression [%s]", expr);
      }
      else {
        // exception in invoked method
        errMsg = String.format("Caught an exception in evaluating expression [%s]", expr);
      }

      logStackTrace(errMsg, e);
    }
    if (throwExceptionOnFailure) {
      throw new XWorkException(e, errMsg);
    }
    return null;
  }
```
And 
```
@Override
  protected Object handleOtherException(String expr, boolean throwExceptionOnFailure, Exception e) {
    if (shouldLogWarning()) {
      if (e.getCause() == null) {
        logStackTrace(String.format("Could not evaluate expression [%s]", expr), e);
      }
      else {
        logStackTrace(String.format("Caught an exception in evaluating expression [%s]", expr), e);
      }
    }
    if (throwExceptionOnFailure) {
      throw new XWorkException(e);
    }
    return findInContext(expr);
  }
```

Reply:
I'm still not sure if we should remove `elseif (reason != null)`. In my view it changes the flow too much (validity and performance), and it's irrelevant to the issue.

Reply:
We should because with that following test fails. I tested and even Struts 2.5 (5/5/2016) with no change also fails on this test: https://github.com/apache/struts/blob/a50af87644320a3ca85283ccab7bc72822cb230b/core/src/test/java/com/opensymphony/xwork2/ognl/OgnlValueStackTest.java#L982
but this is and was a wrong behavior.